### PR TITLE
Support PHP 7.3 7.4 8.0

### PR DESCRIPTION
**Pull request by phil-davis**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no (dropping support for old PHP versions is not considered a BC break - see comment below)
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

Issue #11

### CODE REVIEW

**Comment by Ocramius** [comment]
> BC Break | yes (drops support for lots of old PHP versions)

Just clarifying this misconception: dropping support for older versions is not a BC break as long as `composer.json` ranges state it clearly, since we didn't change any of the API of this package.

https://semver.org/spec/v2.0.0.html#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api

**Review by Ocramius — CHANGES_REQUESTED** [review]
Some of the tests need to be upgraded, as highlighted by CI output:

```
PHP Fatal error:  Declaration of LaminasTest\Ldap\BindTest::setUp() must be compatible with PHPUnit\Framework\TestCase::setUp(): void in /home/travis/build/laminas/laminas-ldap/test/BindTest.php on line 333
```

**Comment by phil-davis** [comment]
I am looking... 
Changing to Ubuntu 20.4 for all PHP versions is not working (PHP 7.3 does not install - I will use Ubuntu 18.04 for that)

And there are a whole hep of unit test fails in https://travis-ci.com/github/laminas/laminas-ldap/jobs/469510058

Time to run the unit tests locally and see what needs changing...

**Comment by phil-davis** [comment]
@Ocramius there are some uses of `assertAttributeEquals()` in  the tests. Those look inside objects for the values of private properties - which his "strictly naughty". `assertAttributeEquals()` has been removed from phpunit9.

The places where it is used in the tests are doing things like `$ldap->connect();` and then "looking inside" to see if `connectString` has the expected value. That string is used in some exception messages, e.g. when the LDAP server goes down. I guess it was harder to test some of those exceptions (hard to make the problem happen in a real controlled way)

What do you want to do about those `assertAttributeEquals()` ?

Edit: I have refactored to remove `assertAttributeEquals` and these tests are now passing.

```
There were 5 errors:

1) LaminasTest\Ldap\ConnectTest::testConnectionWithoutPortInOptionsArray with data set #0 ('localhost', false, 'ldap://localhost:389')
Error: Call to undefined method LaminasTest\Ldap\ConnectTest::assertAttributeEquals()
/home/travis/build/laminas/laminas-ldap/test/ConnectTest.php:281

2) LaminasTest\Ldap\ConnectTest::testConnectionWithoutPortInOptionsArray with data set #1 ('localhost', true, 'ldaps://localhost:636')
Error: Call to undefined method LaminasTest\Ldap\ConnectTest::assertAttributeEquals()
/home/travis/build/laminas/laminas-ldap/test/ConnectTest.php:281

3) LaminasTest\Ldap\SortTest::testSettingCallable
Error: Call to undefined method LaminasTest\Ldap\SortTest::assertAttributeEquals()
/home/travis/build/laminas/laminas-ldap/test/SortTest.php:44

4) LaminasTest\Ldap\SortTest::testSorting
Error: Call to undefined method LaminasTest\Ldap\SortTest::assertAttributeEquals()
/home/travis/build/laminas/laminas-ldap/test/SortTest.php:65

5) LaminasTest\Ldap\SortTest::testCustomSorting
Error: Call to undefined method LaminasTest\Ldap\SortTest::assertAttributeEquals()
/home/travis/build/laminas/laminas-ldap/test/SortTest.php:122
```

**Comment by phil-davis** [comment]
https://travis-ci.com/github/laminas/laminas-ldap/jobs/469590024
```
There were 29 errors:

1) LaminasTest\Ldap\ConnectTest::testConnectionWithoutPortInOptionsArray with data set #0 ('localhost', false, 'ldap://localhost:389')

Error: Call to undefined method LaminasTest\Ldap\ConnectTest::assertAttributeEquals()

/home/travis/build/laminas/laminas-ldap/test/ConnectTest.php:281

2) LaminasTest\Ldap\ConnectTest::testConnectionWithoutPortInOptionsArray with data set #1 ('localhost', true, 'ldaps://localhost:636')

Error: Call to undefined method LaminasTest\Ldap\ConnectTest::assertAttributeEquals()

/home/travis/build/laminas/laminas-ldap/test/ConnectTest.php:281

3) LaminasTest\Ldap\CopyRenameTest::testRenameSourceNotExists

Laminas\Ldap\Exception\LdapException: 0x20 (No such object): renaming ou=DoesNotExist,ou=test,dc=example,dc=com to ou=NewTest,ou=test,dc=example,dc=com

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1770

/home/travis/build/laminas/laminas-ldap/test/CopyRenameTest.php:144

4) LaminasTest\Ldap\CopyRenameTest::testRenameTargetExists

Laminas\Ldap\Exception\LdapException: 0x44 (Already exists): renaming ou=OrgTest,ou=test,dc=example,dc=com to ou=Test1,ou=test,dc=example,dc=com

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1770

/home/travis/build/laminas/laminas-ldap/test/CopyRenameTest.php:152

5) LaminasTest\Ldap\CopyRenameTest::testRenameTargetParentNotExists

Laminas\Ldap\Exception\LdapException: 0x20 (No such object; new superior not found): renaming ou=OrgTest,ou=test,dc=example,dc=com to ou=Test1,ou=ParentDoesNotExist,ou=test,dc=example,dc=com

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1770

/home/travis/build/laminas/laminas-ldap/test/CopyRenameTest.php:160

6) LaminasTest\Ldap\CopyRenameTest::testRenameEmulationSourceNotExists

Laminas\Ldap\Exception\LdapException: 0x20 (No such object): searching: (objectClass=*)

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1121

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1283

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1823

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1776

/home/travis/build/laminas/laminas-ldap/test/CopyRenameTest.php:168

7) LaminasTest\Ldap\CopyRenameTest::testRenameEmulationTargetExists

Laminas\Ldap\Exception\LdapException: 0x44 (Already exists): adding: ou=Test1,ou=test,dc=example,dc=com

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1388

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1830

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1776

/home/travis/build/laminas/laminas-ldap/test/CopyRenameTest.php:176

8) LaminasTest\Ldap\CopyRenameTest::testRenameEmulationTargetParentNotExists

Laminas\Ldap\Exception\LdapException: 0x20 (No such object): adding: ou=Test1,ou=ParentDoesNotExist,ou=test,dc=example,dc=com

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1388

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1830

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1776

/home/travis/build/laminas/laminas-ldap/test/CopyRenameTest.php:188

9) LaminasTest\Ldap\CrudTest::testIllegalAdd

Laminas\Ldap\Exception\LdapException: 0x20 (No such object): adding: ou=TestCreated,ou=Node2,ou=test,dc=example,dc=com

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1388

/home/travis/build/laminas/laminas-ldap/test/CrudTest.php:74

10) LaminasTest\Ldap\CrudTest::testIllegalDelete

Laminas\Ldap\Exception\LdapException: 0x20 (No such object): deleting: ou=TestCreated,ou=test,dc=example,dc=com

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1495

/home/travis/build/laminas/laminas-ldap/test/CrudTest.php:117

11) LaminasTest\Ldap\CrudTest::testPrepareLDAPEntryArrayArrayData

Laminas\Ldap\Exception\InvalidArgumentException: Only scalar values allowed in LDAP data

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1314

/home/travis/build/laminas/laminas-ldap/test/CrudTest.php:241

12) LaminasTest\Ldap\CrudTest::testPrepareLDAPEntryArrayObjectData

Laminas\Ldap\Exception\InvalidArgumentException: Only scalar values allowed in LDAP data

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1314

/home/travis/build/laminas/laminas-ldap/test/CrudTest.php:253

13) LaminasTest\Ldap\Node\OnlineTest::testLoadFromLDAPIllegalEntry

Laminas\Ldap\Exception\LdapException: 0x20 (No such object): searching: (objectClass=*)

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1121

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1283

/home/travis/build/laminas/laminas-ldap/src/Node.php:271

/home/travis/build/laminas/laminas-ldap/test/Node/OnlineTest.php:84

14) LaminasTest\Ldap\Node\OnlineTest::testAttachToInvalidLDAP

Laminas\Ldap\Exception\LdapException: 0x50: LDAP connection is not responsible for given node.

/home/travis/build/laminas/laminas-ldap/src/Node.php:147

/home/travis/build/laminas/laminas-ldap/test/Node/OnlineTest.php:123

15) LaminasTest\Ldap\Node\OnlineTest::testGetIllegalNode

Laminas\Ldap\Exception\LdapException: 0x20 (No such object): searching: (objectClass=*)

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1121

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1283

/home/travis/build/laminas/laminas-ldap/src/Node.php:271

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1854

/home/travis/build/laminas/laminas-ldap/test/Node/OnlineTest.php:179

16) LaminasTest\Ldap\Node\OnlineTest::testGetNonexistentParent

Laminas\Ldap\Exception\LdapException: 0x50: LDAP connection is not responsible for given node.

/home/travis/build/laminas/laminas-ldap/src/Node.php:147

/home/travis/build/laminas/laminas-ldap/src/Node.php:87

/home/travis/build/laminas/laminas-ldap/src/Node.php:275

/home/travis/build/laminas/laminas-ldap/src/Node.php:1053

/home/travis/build/laminas/laminas-ldap/test/Node/OnlineTest.php:267

17) LaminasTest\Ldap\Node\RootDseTest::testSetterWillThrowException

Laminas\Ldap\Exception\BadMethodCallException: 

/home/travis/build/laminas/laminas-ldap/src/Node/AbstractNode.php:349

/home/travis/build/laminas/laminas-ldap/test/Node/RootDseTest.php:129

18) LaminasTest\Ldap\Node\RootDseTest::testOffsetSetWillThrowException

Laminas\Ldap\Exception\BadMethodCallException: 

/home/travis/build/laminas/laminas-ldap/src/Node/AbstractNode.php:408

/home/travis/build/laminas/laminas-ldap/test/Node/RootDseTest.php:138

19) LaminasTest\Ldap\Node\RootDseTest::testUnsetterWillThrowException

Laminas\Ldap\Exception\BadMethodCallException: 

/home/travis/build/laminas/laminas-ldap/src/Node/AbstractNode.php:378

/home/travis/build/laminas/laminas-ldap/test/Node/RootDseTest.php:147

20) LaminasTest\Ldap\Node\RootDseTest::testOffsetUnsetWillThrowException

Laminas\Ldap\Exception\BadMethodCallException: 

/home/travis/build/laminas/laminas-ldap/src/Node/AbstractNode.php:439

/home/travis/build/laminas/laminas-ldap/test/Node/RootDseTest.php:156

21) LaminasTest\Ldap\Node\SchemaTest::testSetterWillThrowException

Laminas\Ldap\Exception\BadMethodCallException: 

/home/travis/build/laminas/laminas-ldap/src/Node/AbstractNode.php:349

/home/travis/build/laminas/laminas-ldap/test/Node/SchemaTest.php:67

22) LaminasTest\Ldap\Node\SchemaTest::testOffsetSetWillThrowException

Laminas\Ldap\Exception\BadMethodCallException: 

/home/travis/build/laminas/laminas-ldap/src/Node/AbstractNode.php:408

/home/travis/build/laminas/laminas-ldap/test/Node/SchemaTest.php:75

23) LaminasTest\Ldap\Node\SchemaTest::testUnsetterWillThrowException

Laminas\Ldap\Exception\BadMethodCallException: 

/home/travis/build/laminas/laminas-ldap/src/Node/AbstractNode.php:378

/home/travis/build/laminas/laminas-ldap/test/Node/SchemaTest.php:83

24) LaminasTest\Ldap\Node\SchemaTest::testOffsetUnsetWillThrowException

Laminas\Ldap\Exception\BadMethodCallException: 

/home/travis/build/laminas/laminas-ldap/src/Node/AbstractNode.php:439

/home/travis/build/laminas/laminas-ldap/test/Node/SchemaTest.php:91

25) LaminasTest\Ldap\SearchTest::testGetSingleIllegalEntryWithException

Laminas\Ldap\Exception\LdapException: 0x20 (No such object): searching: (objectClass=*)

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1121

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1283

/home/travis/build/laminas/laminas-ldap/test/SearchTest.php:56

26) LaminasTest\Ldap\SearchTest::testIllegalSearch

Laminas\Ldap\Exception\LdapException: 0x20 (No such object): searching: (objectClass=account)

/home/travis/build/laminas/laminas-ldap/src/Ldap.php:1121

/home/travis/build/laminas/laminas-ldap/test/SearchTest.php:136

27) LaminasTest\Ldap\SortTest::testSettingCallable

Error: Call to undefined method LaminasTest\Ldap\SortTest::assertAttributeEquals()

/home/travis/build/laminas/laminas-ldap/test/SortTest.php:44

28) LaminasTest\Ldap\SortTest::testSorting

Error: Call to undefined method LaminasTest\Ldap\SortTest::assertAttributeEquals()

/home/travis/build/laminas/laminas-ldap/test/SortTest.php:65

29) LaminasTest\Ldap\SortTest::testCustomSorting

Error: Call to undefined method LaminasTest\Ldap\SortTest::assertAttributeEquals()

/home/travis/build/laminas/laminas-ldap/test/SortTest.php:122

ERRORS!

Tests: 522, Assertions: 1372, Errors: 29, Skipped: 5.
```

I have refactored all the "easy" stuff related to phpunit9.

Tomorrow or at the weekend I can look at the above errors from the PHP 7.4 job. There should be some "easy" way to fix those, because PHP 7.4 passes in the main branch, so the actual code must be OK - it should just be test refactoring.

After that I will look at the PHP 8.0 test results. That has more erros reported. So there might be actual PHP 8.0-related refactoring to do.

**Comment by phil-davis** [comment]
The PHP 8.0 tests have more errors: https://travis-ci.com/github/laminas/laminas-ldap/jobs/469596012

I will look tomorrow or at the weekend. There is probably real refactoring needed for that.

**Comment by phil-davis** [comment]
Tests are passing for PHP 7.3 and 7.4 with phpunit9.
Next is to look at the unit test fails with PHP 8.0 and see what "real"  changes are needed.

@Ocramius we could make the changes so far in this PR, and it would drop all the old PHP versions and just support PHP 7.3 and 7.4, and use phpunit9 for testing. That would provide a cleaned-up point before moving forward to also support PHP 8.0.

I made PR #14 in case you are happy to do that.

**Comment by phil-davis** [comment]
I added a commit to require the various current versions of dependencies that already have support specified similar to "^7.3 || ^8.0" - IMO we may as well clean up references to old versions of dependencies that actually [will|should] never be selected any more.

**Comment by phil-davis** [comment]
Note: `"laminas/laminas-coding-standard": "~1.0.0"`

This gets us an old version of `php_codesniffer` which has ancient code:
```
$ composer cs-check
> phpcs
PHP Deprecated:  Array and string offset access syntax with curly braces is deprecated in /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/CodeSniffer.php on line 2028
PHP Stack trace:
PHP   1. {main}() /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/scripts/phpcs:0
PHP   2. include_once() /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/scripts/phpcs:19
PHP Deprecated:  Array and string offset access syntax with curly braces is deprecated in /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/CodeSniffer.php on line 2033
PHP Stack trace:
PHP   1. {main}() /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/scripts/phpcs:0
PHP   2. include_once() /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/scripts/phpcs:19
PHP Deprecated:  Array and string offset access syntax with curly braces is deprecated in /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/CodeSniffer.php on line 2033
PHP Stack trace:
PHP   1. {main}() /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/scripts/phpcs:0
PHP   2. include_once() /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/scripts/phpcs:19
PHP Deprecated:  Array and string offset access syntax with curly braces is deprecated in /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/CodeSniffer.php on line 2079
PHP Stack trace:
PHP   1. {main}() /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/scripts/phpcs:0
PHP   2. include_once() /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/scripts/phpcs:19
PHP Deprecated:  Array and string offset access syntax with curly braces is deprecated in /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/CodeSniffer.php on line 2079
PHP Stack trace:
PHP   1. {main}() /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/scripts/phpcs:0
PHP   2. include_once() /home/phil/git/laminas/laminas-ldap/vendor/squizlabs/php_codesniffer/scripts/phpcs:19
............................................................ 60 / 84 (71%)
........................

Time: 1.67 secs; Memory: 26Mb
```

`CodeSnifffer.php` has ancient code at line 20333:
```
                    if (strtoupper($string{$i}) === $string{$i}) {
```

To get a newer `phpcs` we need to move to major version 2.* of `laminas/laminas-coding-standard` But if I do that then it finds a whole lot of things to report. So IMO that is something for a separate PR.

**Comment by phil-davis** [comment]
In PHP 8.0 various "ldap*" functions now complain if you do not pass them a valid LDAP resource from the initial connect. unbind also complains if you try to unbind a resource that never got bound properly, and if you try to unbind a bind that had a null username.

For example:
```
1) LaminasTest\Ldap\OfflineReconnectTest::testRemovingAttributesFails
Failed asserting that exception of type "TypeError" matches expected exception "Laminas\Ldap\Exception\LdapException". Message was: "ldap_error(): Argument #1 ($ldap) must be of type resource, null given" at
/home/phil/git/laminas/laminas-ldap/src/Ldap.php:180
/home/phil/git/laminas/laminas-ldap/src/Exception/LdapException.php:108
/home/phil/git/laminas/laminas-ldap/src/Ldap.php:1650
/home/phil/git/laminas/laminas-ldap/test/OfflineTest.php:186
```

So I have added a commit that keeps track of what bind has been done, and of an unbind is actually needed. And double-checks in the error processing to only try to call the various "get last LDAP error" functions if a connect has actually been done.

That gets unit tests passing in PHP 8.0, without breaking anything for PHP 7.3 and 7.4.

@Ocramius please review again.

File: src/Collection/DefaultIterator.php
Comment (phil-davis) [inline src/Collection/DefaultIterator.php]: This is an accident - I will remove it.

**Comment by Ocramius** [comment]
CI seems to be grren: is this still a draft now?

**Comment by phil-davis** [comment]
I changed it to "open/ready" I requested your review.
"marked this pull request as ready for review 17 hours ago"
"requested a review from Ocramius 17 hours ago"

Please review and merge if OK.

File: src/Ldap.php
Comment (heiglandreas) [inline src/Ldap.php]: Can we init the `$estr1` here with `getLastError: could not call ldap_error ...` and reassign the result from `ldap_error` in case we have a proper resource? That would eliminate the `else`.

File: src/Ldap.php
Comment (heiglandreas) [inline src/Ldap.php]: Can we init the `$estr1` here with `getLastError: could not call ldap_error ...` and reassign the result from `ldap_error` in case we have a proper resource? That would eliminate the `else`.

File: src/Ldap.php
Comment (heiglandreas) [inline src/Ldap.php]: When does a `bind` not need `unbind`? If I got that correclty this field declares whether an anonymous bind has been done or not. So don't we already have this infoormation in the field `boundUser`?  If `$boundUser` is `NULL` we do not need an unbind.

File: src/Ldap.php
Comment (heiglandreas) [inline src/Ldap.php]:
```suggestion
        if (is_resource($this->resource) && (is_string($this->boundUser))) {
```

File: src/Ldap.php
Comment (heiglandreas) [inline src/Ldap.php]:
```suggestion
```

File: src/Ldap.php
Comment (heiglandreas) [inline src/Ldap.php]:
```suggestion
```

File: src/Ldap.php
Comment (heiglandreas) [inline src/Ldap.php]:
```suggestion
```

File: src/Ldap.php
Comment (heiglandreas) [inline src/Ldap.php]:
```suggestion
```

**Review by heiglandreas — CHANGES_REQUESTED** [review]
Thanks @phil-davis for this awesome work. I do though have some small things that might reduce a bit of the added complexity by reusing already existing informations.

File: src/Ldap.php
Comment (phil-davis) [reply src/Ldap.php]: I got rid of this var and used your suggestion. It ends up being equivalent code.

File: src/Ldap.php
Comment (heiglandreas) [inline src/Ldap.php]:
```suggestion
            }
            $this->boundUser = false;
```

As the `if` condition will **always** return we can skip the `else` here.

File: src/Ldap.php
Comment (heiglandreas) [inline src/Ldap.php]:
```suggestion
        $ret = false;
        if (is_resource($this->resource)) {
            $ret = ldap_get_option($this->resource, LDAP_OPT_ERROR_NUMBER, $err);
        }
```

We can also preset the `$ret` variable here with `false` and eliminate the `else`

**Review by heiglandreas — CHANGES_REQUESTED** [review]

File: src/Ldap.php
Comment (phil-davis) [reply src/Ldap.php]: Will do. Personally I feel that explicitly writing the `if` and `else` cases is more readable. But it is just coding-style.

File: src/Ldap.php
Comment (phil-davis) [reply src/Ldap.php]: Done

File: .travis.yml
Comment (froschdesign) [inline .travis.yml]: Please revert this change and it to the related PHP version. You can compare with the Travis configuration of laminas-validator: https://github.com/laminas/laminas-validator/blob/23623b4756eab850829e5b5f6205eeaf1303e959/.travis.yml#L7-L38

**Comment by phil-davis** [comment]
@heiglandreas let me know if you want me to squash the long list of commits.

**Review by heiglandreas — APPROVED** [review]

File: composer.json
Comment (froschdesign) [inline composer.json]: Please use the version constrain like described in the related issue report #11:
```suggestion
        "php": "^7.3 || ~8.0.0",
```

**Comment by heiglandreas** [comment]
From my POV the list of commits is absolutely OK. No need to squash them.

File: phpunit.xml.dist
Comment (froschdesign) [inline phpunit.xml.dist]: The version of PHPUnit is controlled via `composer.json` and the schema is installed locally:
```suggestion
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd" bootstrap="test/bootstrap.php" colors="true">
```

File: src/Collection/DefaultIterator.php
Comment (froschdesign) [inline src/Collection/DefaultIterator.php]: Please revert this change because extending the API to test internal properties is not an option. Use reflection, an anonymous class or better find a test scenario that do not need access to internal operations.

**Review by froschdesign — CHANGES_REQUESTED** [review]

File: .travis.yml
Comment (phil-davis) [reply .travis.yml]: done - I kept `--no-interaction` in the PHP 8.0 value also, that seems "a good thing" to make sure that composer does not accidentally ask questions.

File: composer.json
Comment (phil-davis) [reply composer.json]: Done. For my education, why not do it like is done for `^7.3` ?
What difference does it make?

File: phpunit.xml.dist
Comment (phil-davis) [reply phpunit.xml.dist]: This was generated by `phpunit` with the `--migrate-configuration` command.

File: src/Collection/DefaultIterator.php
Comment (phil-davis) [reply src/Collection/DefaultIterator.php]: @froschdesign I thought for a while about the related tests, and it will be painful to find a way to "replace" this test with "something useful". I can certainly change the tests to use Reflection. I thought that I would not get away with just adding a getter :)

I found https://github.com/laminas/laminas-ldap/pull/13/files#diff-9ad7ec52897b7302bf9c4dd9b3c4c04e903b948aab795ef87c25db6e6db46078R74 sortTest.php testSorting()
and https://github.com/laminas/laminas-ldap/pull/13/files#diff-9ad7ec52897b7302bf9c4dd9b3c4c04e903b948aab795ef87c25db6e6db46078R116 sortTest.php testCustomSorting()

It looks to me that those test cases cover using the default, and a different sorting function.

So actually, why do we need the test cases that are trying to check the internal value of the private variable `sortFunction` ?

Comment (phil-davis) [reply src/Collection/DefaultIterator.php]: I have made the change to use reflection, and I removed this getter.

**Comment by phil-davis** [comment]
I squashed the commits into a more-reasonable set, getting rid of all the little commits from reviewing.
@froschdesign @Ocramius @heiglandreas ready for review again.

File: test/CrudTest.php
Comment (phil-davis) [inline test/CrudTest.php:75]: Note: this was "bonus" code that does not execute, because there is an exception that happens at the line above.

**Comment by phil-davis** [comment]
@Ocramius I made the changes here that you suggested in your review of PR #14 

Since this PR has code ready for PHP 8.0 and passing, maybe this can be merged.

Let me know if you want me to re-create PR #14 with the code from here, minus the PHP 8.0 stuff.

File: .travis.yml
Comment (Ocramius) [inline .travis.yml]: Does installation work without the `--ignore-platform-reqs` now? If so, would drop it before merging

File: test/BindTest.php
Comment (Ocramius) [inline test/BindTest.php]: Note: this kind of assertion will trip static analysis (to be introduced in future).

Is there anything else we can assert upon? For example getting something from `$ldap` and verifying its value, rather than its shape?

File: test/CrudTest.php
Comment (Ocramius) [reply test/CrudTest.php:75]: Good catch! 👍

File: test/CrudTest.php
Comment (Ocramius) [inline test/CrudTest.php]: Also here: this will fail at static analysis (coming up in future).

Anything valuable we can assert upon at all?

**Review by Ocramius — APPROVED** [review]
> Let me know if you want me to re-create PR #14 with the code from here, minus the PHP 8.0 stuff.

I don't mind either way: IMO if this already includes #14, we can close that one, but really up to how you prefer working.

As for this patch, I still found a couple things to be checked, but we're almost there 👍

File: test/CrudTest.php
Comment (phil-davis) [reply test/CrudTest.php]: I will be offline for a couple of hours. I will look at these tonight. They are all only in test code and Travis script so do not effect "real release". If you want to move this along and get a "real release" then feel free to merge this and I can sort this stuff out in another PR. Or wait...

(This PR seems to be dragging on "forever" - each time I address review comments, then more come from the next review)

Comment (Ocramius) [reply test/CrudTest.php]: > (This PR seems to be dragging on "forever" - each time I address review comments, then more come from the next review)

That's because as the diff becomes clearer, so do the issues with it.

It's always push and pull though: you have to consider that once this is "done", it is work on the maintainers to chase any problems with it long-term, so we do indeed pay a lot of attention to detail too.

Comment (phil-davis) [reply test/CrudTest.php]: I changed it to check the value of `getLastErrorCode()`

File: test/BindTest.php
Comment (phil-davis) [reply test/BindTest.php]: I changed it to check the value of `getLastErrorCode()`

File: .travis.yml
Comment (phil-davis) [reply .travis.yml]: I removed this, and CI still passes - good. So I don't know why the issue had an action point "Modify .travis.yml to ignore platform requirements when installing composer dependencies" - I guess that point is out-of-date.

**Comment by phil-davis** [comment]
@Ocramius @froschdesign ready for review again

**Review by Ocramius — APPROVED** [review]
Excellent, LGTM! 

Merging and releasing on this end, thanks @phil-davis!

**Pull request merged**
